File: web/src/krate/ktab/po.ts
import type { TID } from "@/lib/id_util";
import type { DbText, Varchar } from "@/lib/types";

export type KTabColumnStoreKind = "str" | "i64" | "f64" | "date" | "blob";

export type KTabColumnViewKind =
  | "text"
  | "string"
  | "number"
  | "integer"
  | "date"
  | "datetime"
  | "decimal"
  | "progress"
  | "checkbox"
  | "multi_select"
  | "image";

// To use Decimal to avoid any lost.
type Decimal = string;

export const ktabViewToStoreKind = (
  kind: KTabColumnViewKind,
): KTabColumnStoreKind => {
  switch (kind.toLowerCase()) {
    case "text":
    case "string":
      return "str";
    case "number":
    case "integer":
      return "i64";
    case "float":
    case "decimal":
      return "f64";
    case "date":
    case "datetime":
      return "date";
    case "image":
      return "str";
    case "blob":
      return "blob";
    case "bool":
    case "checkbox":
      return "i64";
    default:
      throw new Error(`unable to map ${kind}`);
  }
};

export type KTabColumnMeta = {
  idx: TID;
  name: string;
  comment: string;
  store_kind: KTabColumnStoreKind;
  view_kind: KTabColumnViewKind;
  required: boolean;
  order_by: number;
};

export type KTabMeta = {
  otid: TID;
  columns: Record<string, KTabColumnMeta>;
  table_name: Varchar<300>;
  table_comment: Varchar<1000>;
  update_time?: Date;
  real_table: boolean;
  tid: TID;
};

export type KTabDisplayAsGroup = {
  group: string;
  items: { viewKind: KTabColumnViewKind; label: string }[];
};

export const KTAB_DISPLAY_AS_MENU: KTabDisplayAsGroup[] = [
  {
    group: "文本",
    items: [
      { viewKind: "text", label: "Text" },
      { viewKind: "string", label: "String" },
    ],
  },
  {
    group: "数字",
    items: [
      { viewKind: "number", label: "Number" },
      { viewKind: "integer", label: "Integer" },
      { viewKind: "decimal", label: "Decimal" },
    ],
  },
  {
    group: "日期",
    items: [
      { viewKind: "date", label: "日期" },
      { viewKind: "datetime", label: "日期时间" },
    ],
  },
  {
    group: "其他",
    items: [
      { viewKind: "progress", label: "进度" },
      { viewKind: "checkbox", label: "Checkbox" },
      { viewKind: "multi_select", label: "枚举" },
      { viewKind: "image", label: "图片" },
    ],
  },
];

export type KTabCell = {
  table_id: number;
  col_tid: number;
  row_tid: number;
  cell_data: KTabCellData;
};

export type KTabCellData = { Date: Date } | { Text: string };

export type KTabCellText = {
  table_otid: TID;
  col_otid: TID;
  row_otid: TID;
  tid: TID;
  cell_data: DbText;
};

export type KTabCellDecimal = {
  table_otid: TID;
  col_otid: TID;
  row_otid: TID;
  tid: TID;
  cell_data: Decimal;
};

export type KTabCellDate = {
  table_otid: TID;
  col_otid: TID;
  row_otid: TID;
  tid: TID;
  cell_data: Date;
};

File: backend/core/src/krate/ktab/po.rs
use std::collections::HashMap;

use chin_sql::str_type::Text;
use chin_sql::GenerateTableSchema;
use chin_sql::{str_type::Varchar, time_type::TID};
use chrono::{DateTime, FixedOffset};
use serde::{Deserialize, Serialize};

use crate::model::decimal::Decimal;

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum KTabColumnStoreKind {
    Decimal,
    Str,
    Date,
    // Blob,
}

pub type KTabColumnViewKind = String;
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct KTabColumnMeta {
    pub idx: TID,
    pub name: String,
    pub comment: String,
    pub store_kind: KTabColumnStoreKind,
    pub view_kind: KTabColumnViewKind,
    pub required: bool,
    pub order_by: i32,
}

fn map_to_sql(value: HashMap<String, KTabColumnMeta>) -> String {
    // Only serde_json do not fail.
    serde_json::to_string(&value).unwrap()
}

#[derive(Clone, Debug, Serialize, Deserialize, GenerateTableSchema)]
pub struct KTabMeta {
    #[gts_primary]
    #[gts_type = "i64"]
    pub otid: TID,

    #[gts_type = "Text"]
    #[gts_tosql = "map_to_sql"]
    pub columns: HashMap<String, KTabColumnMeta>,
    pub table_name: Varchar<300>,
    pub table_comment: Varchar<1000>,
    pub update_time: Option<DateTime<FixedOffset>>,
    pub real_table: bool,

    #[gts_unique]
    #[gts_type = "i64"]
    pub tid: TID,
}

macro_rules! type_table {
    ($sname:tt, $data_type:ty $(, #[$attr:meta])*) => {
        #[derive(Clone, Debug, Serialize, Deserialize, GenerateTableSchema)]
        pub struct $sname {
            #[gts_primary]
            #[gts_type = "i64"]
            pub table_otid: TID,

            #[gts_primary]
            #[gts_type = "i64"]
            pub col_otid: TID, // actually is the insert time(unix timestamp), so it is easy for data merge

            #[gts_primary]
            #[gts_type = "i64"]
            pub row_otid: TID, // actually is the insert time(unix timestamp), so it is easy for data merge

            #[gts_unique]
            #[gts_type = "i64"]
            pub tid: TID,

            $(#[$attr])*
            pub cell_data: $data_type
        }
    }
}

type_table!(KTabCellText, Text);
type_table!(KTabCellDecimal, Decimal, #[gts_type = "Text"]);
type_table!(KTabCellDate, DateTime<FixedOffset>);

#[cfg(test)]
mod ktab_column_type_test {
    #[test]
    fn test() {
        use crate::krate::ktab::po::KTabColumnStoreKind;

        println!("{:?}", serde_json::to_string(&KTabColumnStoreKind::Str));
    }
}
